### PR TITLE
maint: add instrumentation for GoRedisPubSub

### PR DESCRIPTION
**Pull request by VinozzZ**
## Which problem is this PR solving?

Add instrumentation for GoRedisPubSub to gain more insight on how the redis pubsub system works

## Short description of the changes

- add tracer to `GoRedisPubSub`
- fix tests
- define a type for the pubsub subscription callback function

### CODE REVIEW

File: pubsub/pubsub_goredis.go [hunk 1]
```diff
@@ -25,6 +28,7 @@ type GoRedisPubSub struct {
 	Config  config.Config   `inject:""`
 	Logger  logger.Logger   `inject:""`
 	Metrics metrics.Metrics `inject:"metrics"`
+	Tracer  trace.Tracer    `inject:"tracer"`
```
Comment (kentquirk) [inline pubsub/pubsub_goredis.go:31]: If we don't expect to inject multiple different tracers, we don't need to name them. You could remove the name here, and in main where it's created.

File: pubsub/pubsub_goredis.go [hunk 2]
```diff
@@ -139,8 +153,18 @@ func (ps *GoRedisPubSub) Subscribe(ctx context.Context, topic string, callback f
 				if msg == nil {
 					continue
 				}
+				ctx, span := otelutil.StartSpanMulti(context.Background(), ps.Tracer, "GoRedisPubSub.Receive", map[string]interface{}{
```
Comment (kentquirk) [inline pubsub/pubsub_goredis.go:156]: I'm slightly worried about the performance overhead of creating a context every time. How about creating just one outside the for loop and reusing it for each span?

**Review by kentquirk — APPROVED** [review]

File: pubsub/pubsub_goredis.go [hunk 3]
```diff
@@ -139,8 +154,18 @@ func (ps *GoRedisPubSub) Subscribe(ctx context.Context, topic string, callback f
...
+					"message":            msg.Payload,
+				})
 				ps.Metrics.Count("redis_pubsub_received", 1)
-				go sub.cb(msg.Payload)
+
+				go func(cbCtx context.Context, span trace.Span) {
+					defer span.End()
+
+					sub.cb(cbCtx, msg.Payload)
+				}(receiveCtx, span)
```
Comment (kentquirk) [inline pubsub/pubsub_goredis.go:168]: Why did you pass the context and span as parameters, but not the payload? I was originally thinking you needed none of them, but now, after thinking it through, it's probably a good idea to pass the message payload as well.

Comment (VinozzZ) [reply pubsub/pubsub_goredis.go]: Good catch, I just changed it

**Pull request merged**